dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+    """
+    Apply a generalized ufunc [2]_ to arrays. The function is
+    mapped to the input arguments. The ``signature`` determines
+    if the function consumes or produces core dimensions. The
+    remaining dimensions in given input arrays (``*args``) are
+    considered loop dimensions and are required to broadcast
+    naturally against each other.
+
+    In other terms, this function is like np.vectorize, but for
+    the blocks of dask arrays. If the function itself shall also
+    be vectorized, ``vectorize=True`` can be used for convenience.
+
+    Parameters
+    ----------
+    func : callable
+        Function to call like ``func(*args, **kwargs)`` on input arrays
+        (``*args``) that returns an array or tuple of arrays. If multiple
+        arguments with non-matching dimensions are supplied, this function is
+        expected to vectorize (broadcast) over axes of positional arguments in
+        the style of NumPy universal functions [1]_ (if this is not the case,
+        set ``vectorize=True``).
+    signature: String
+        Specifies what core dimensions are consumed and produced by ``func``.
+        According to the specification of numpy.gufunc signature [2]_
+    output_dtypes : dtype or list of dtypes
+        dtype or list of output dtypes.
+    output_sizes : dict, optional
+        Optional mapping from dimension names to sizes for outputs. Only used if
+        new core dimensions (not found on inputs) appear on outputs.
+    vectorize: bool
+        If set to ``True``, ``np.vectorize`` is applied to ``func`` for
+        convenience. Defaults to ``False``.
+
+    Returns
+    -------
+    Single dask.array.Array or tuple of dask.array.Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    ... import numpy as np
+    ... def stats(x):
+    ...     return np.mean(x, axis=-1), np.std(x, axis=-1)
+    ... a = da.random.normal(size=(10,20,30), chunks=5)
+    ... mean, std = da.apply_gufunc(stats, "(i)->(),()", a, output_dtypes=2*(a.dtype,))
+    ... mean.compute()  # doctest: +SKIP
+
+    >>> import dask.array as da
+    ... import numpy as np
+    ... def outer_product(x, y):
+    ...     return np.einsum("...i,...j->...ij", x, y)
+    ... a = da.random.normal(size=(   20,30), chunks=5)
+    ... b = da.random.normal(size=(10, 1,40), chunks=10)
+    ... c = da.apply_gufunc(outer_product, "(i),(j)->(i,j)", a, b, output_dtypes=a.dtype)
+    ... c.compute()  # doctest: +SKIP
+
+    References
+    ----------
+    .. [1] http://docs.scipy.org/doc/numpy/reference/ufuncs.html
+    .. [2] http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html
+    """
+    output_dtypes = kwargs.pop("output_dtypes", None)
+    output_sizes = kwargs.pop("output_sizes", None)
+    vectorize = kwargs.pop("vectorize", None)
+    concatenate = kwargs.pop("concatenate", True)
+    if output_dtypes is None:
+        raise ValueError("Must specify `output_dtypes` of output array(s)")
```

Comment:
Typically we run the function on a small or empty array to see if that can produce the dtype for us.  See `dask/array/core.py::apply_infer_dtype`